net/xia/ppal_ether/main.c

```diff
@@ -162,6 +162,296 @@ static int local_dump_ether(struct fib_xid *fxid, struct fib_xid_table *xtbl,
+	*/
+	spin_lock(&einterface->interface_lock);
+	list_del_rcu(&mether->interface_common_addr);
+	spin_unlock(&einterface->interface_lock);
+
+	ether_interface_put(einterface);
+}
+
+/* ETHER main table operations. */
+static int main_newroute(struct xip_ppal_ctx *ctx, struct fib_xid_table *xtbl,
+			 struct xia_fib_config *cfg)
+{
+	struct xip_deferred_negdep_flush *dnf;
+	struct fib_xid_ether_main *mether;
+	struct net_device *out_interface;
+	struct fib_xid *cur_fxid;
+	u32 nl_flags, bucket;
+	const char *id;
+	int rc;
+
+	/* Check for errors in cfg. */
+	if (!cfg->xfc_dst || !cfg->xfc_odev || !cfg->xfc_lladdr ||
+	    cfg->xfc_lladdr_len != cfg->xfc_odev->addr_len)
+		return -EINVAL;
+
+	out_interface = cfg->xfc_odev;
+	nl_flags      = cfg->xfc_nlflags;
+	id            = cfg->xfc_dst->xid_id;
+
+	if (!(out_interface->flags & IFF_UP) ||
+	    (out_interface->flags & IFF_LOOPBACK))
+		return -EINVAL;
+
+	cur_fxid = ether_rt_iops->fxid_find_lock(&bucket, xtbl, id);
+	if (cur_fxid) {
+		/* Found a matching fxid. */
+		if (cur_fxid->fx_table_id != XRTABLE_MAIN_INDEX) {
+			rc = -EINVAL;
+			goto unlock_bucket;
+		}
+		mether = fxid_mether(cur_fxid);
+
+		if (!memcmp(mether->ha, cfg->xfc_lladdr, cfg->xfc_lladdr_len) &&
+		    (mether->host_interface == out_interface)) {
+			/* Exact duplicate entry request. */
+			if ((nl_flags & NLM_F_EXCL) ||
+			    !(nl_flags & NLM_F_REPLACE)) {
+				rc = -EEXIST;
+				goto unlock_bucket;
+			}
+			rc = 0;
+			goto unlock_bucket;
+		}
+
+		if (!(nl_flags & NLM_F_REPLACE)) {
+			rc = -EEXIST;
+			goto unlock_bucket;
+		}
+		mether->host_interface = out_interface;
+		attach_neigh_to_interface(mether);
+		ether_rt_iops->fib_unlock(xtbl, &bucket);
+		return 0;
+	}
+
+	if (!(nl_flags & NLM_F_CREATE)) {
+		rc = -ENOENT;
+		goto unlock_bucket;
+	}
+
+	dnf = fib_alloc_dnf(GFP_KERNEL);
+	if (!dnf) {
+		rc = -ENOMEM;
+		goto unlock_bucket;
+	}
+
+	mether = ether_rt_iops->fxid_ppal_alloc(sizeof(*mether), GFP_KERNEL);
+	if (!mether) {
+		rc = -ENOMEM;
+		goto def_upd;
+	}
+	fxid_init(xtbl, &mether->xem_common, id, XRTABLE_MAIN_INDEX, 0);
+	xdst_init_anchor(&mether->xem_anchor);
+	INIT_LIST_HEAD(&mether->interface_common_addr);
+
+	mether->host_interface = out_interface;
+	memmove(mether->ha, cfg->xfc_lladdr, cfg->xfc_lladdr_len);
+	attach_neigh_to_interface(mether);
+
+	WARN_ON(ether_rt_iops->fxid_add_locked(&bucket, xtbl,
+					       &mether->xem_common));
+	ether_rt_iops->fib_unlock(xtbl, &bucket);
+	fib_defer_dnf(dnf, ctx_ether(ctx)->net, XIDTYPE_ETHER);
+	return 0;
+
+def_upd:
+	fib_free_dnf(dnf);
+unlock_bucket:
+	ether_rt_iops->fib_unlock(xtbl, &bucket);
+	return rc;
+}
+
+static int main_delroute(struct xip_ppal_ctx *ctx,
+			 struct fib_xid_table *xtbl, struct xia_fib_config *cfg)
+{
+	struct fib_xid_ether_main *mether;
+	struct net_device *dev;
+	struct fib_xid *fxid;
+	const char *id;
+	u32 bucket;
+	int rc;
+
+	/* Check for errors in cfg. */
+	if (!cfg->xfc_dst || !cfg->xfc_odev || !cfg->xfc_lladdr ||
```

Comment:
You have all the information you need in `xfc_dst`, you don't need `xfc_odev` and `xfc_lladdr`.

This will simplify the xipconf changes.